File: isoforge/diff.py
"""Scene diffing.

Produces RFC-6902 operations between two scenes. Arrays of shapes are matched by `id`
rather than by position, because a positional diff reports "every shape changed" when
one shape is inserted at the front — useless in a tool whose whole job is showing what
the model actually edited.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any


@dataclass(slots=True)
class PatchOp:
    op: str
    path: str
    value: Any = None

    def to_dict(self) -> dict:
        out: dict[str, Any] = {"op": self.op, "path": self.path}
        if self.op in ("add", "replace", "test"):
            out["value"] = self.value
        return out


@dataclass(slots=True)
class Diff:
    ops: list[PatchOp] = field(default_factory=list)

    @property
    def changed(self) -> bool:
        return bool(self.ops)

    @property
    def added(self) -> int:
        return sum(1 for o in self.ops if o.op == "add")

    @property
    def removed(self) -> int:
        return sum(1 for o in self.ops if o.op == "remove")

    @property
    def replaced(self) -> int:
        return sum(1 for o in self.ops if o.op == "replace")

    def to_list(self) -> list[dict]:
        return [o.to_dict() for o in self.ops]


def _same(a: Any, b: Any) -> bool:
    return json.dumps(a, sort_keys=True) == json.dumps(b, sort_keys=True)


def _escape(key: str) -> str:
    return key.replace("~", "~0").replace("/", "~1")


def _keyed_by_id(items: list) -> bool:
    return bool(items) and all(isinstance(i, dict) and isinstance(i.get("id"), str) for i in items)


def _diff_value(path: str, a: Any, b: Any) -> list[PatchOp]:
    if _same(a, b):
        return []

    if isinstance(a, dict) and isinstance(b, dict):
        return _diff_object(path, a, b)

    if isinstance(a, list) and isinstance(b, list) and _keyed_by_id(a) and _keyed_by_id(b):
        return _diff_keyed_array(path, a, b)

    return [PatchOp("replace", path, b)]


def _diff_object(path: str, a: dict, b: dict) -> list[PatchOp]:
    ops: list[PatchOp] = []
    for key in sorted(set(a) | set(b)):  # sorted for deterministic output
        child = f"{path}/{_escape(key)}"
        if key in a and key not in b:
            ops.append(PatchOp("remove", child))
        elif key not in a and key in b:
            ops.append(PatchOp("add", child, b[key]))
        else:
            ops.extend(_diff_value(child, a[key], b[key]))
    return ops


def _diff_keyed_array(path: str, a: list, b: list) -> list[PatchOp]:
    ops: list[PatchOp] = []
    a_index = {item["id"]: item for item in a}
    b_index = {item["id"]: item for item in b}

    # Removals first, in reverse order, so earlier indices stay valid as the patch is
    # applied sequentially.
    for i in range(len(a) - 1, -1, -1):
        if a[i]["id"] not in b_index:
            ops.append(PatchOp("remove", f"{path}/{i}"))

    # Modifications to surviving shapes, addressed by their new position.
    for new_index, item in enumerate(b):
        if (previous := a_index.get(item["id"])) is not None:
            ops.extend(_diff_value(f"{path}/{new_index}", previous, item))

    # Additions last.
    for item in b:
        if item["id"] not in a_index:
            ops.append(PatchOp("add", f"{path}/-", item))

    return ops


def diff_scenes(before: dict, after: dict) -> Diff:
    """Compare two scenes and return the minimal patch between them."""
    return Diff(_diff_value("", before, after))

File: internal/store/diff.go
package store

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// Diff compares two versions of a project's scene and returns an RFC-6902 patch.
func (s *Store) Diff(projectID string, fromVersion, toVersion int) (*DiffResult, error) {
	from, err := s.GetSceneVersionByNumber(projectID, fromVersion)
	if err != nil {
		return nil, fmt.Errorf("from version %d: %w", fromVersion, err)
	}
	to, err := s.GetSceneVersionByNumber(projectID, toVersion)
	if err != nil {
		return nil, fmt.Errorf("to version %d: %w", toVersion, err)
	}

	var a, b any
	if err := json.Unmarshal(from.Scene, &a); err != nil {
		return nil, err
	}
	if err := json.Unmarshal(to.Scene, &b); err != nil {
		return nil, err
	}

	ops := diffValue("", a, b)
	encoded, err := json.Marshal(ops)
	if err != nil {
		return nil, err
	}

	return &DiffResult{
		FromVersion: fromVersion,
		ToVersion:   toVersion,
		FromHash:    from.SceneHash,
		ToHash:      to.SceneHash,
		Changed:     from.SceneHash != to.SceneHash,
		Patch:       encoded,
		Summary:     summarise(ops),
	}, nil
}

// patchOp is one RFC-6902 operation.
type patchOp struct {
	Op    string `json:"op"`
	Path  string `json:"path"`
	Value any    `json:"value,omitempty"`
}

// diffValue produces a minimal patch between two decoded JSON values.
//
// Arrays are diffed by identity where elements carry an "id" (which every shape
// does), rather than positionally. A positional diff would report "every shape
// changed" when one shape is inserted at the front, which is useless in a UI whose
// whole job is showing what the model actually edited.
func diffValue(path string, a, b any) []patchOp {
	if sameJSON(a, b) {
		return nil
	}

	aMap, aIsMap := a.(map[string]any)
	bMap, bIsMap := b.(map[string]any)
	if aIsMap && bIsMap {
		return diffObject(path, aMap, bMap)
	}

	aArr, aIsArr := a.([]any)
	bArr, bIsArr := b.([]any)
	if aIsArr && bIsArr && keyedByID(aArr) && keyedByID(bArr) {
		return diffKeyedArray(path, aArr, bArr)
	}

	return []patchOp{{Op: "replace", Path: path, Value: b}}
}

func diffObject(path string, a, b map[string]any) []patchOp {
	var ops []patchOp

	keys := make([]string, 0, len(a)+len(b))
	seen := map[string]bool{}
	for k := range a {
		keys = append(keys, k)
		seen[k] = true
	}
	for k := range b {
		if !seen[k] {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys) // deterministic patch output

	for _, k := range keys {
		child := path + "/" + escapePointer(k)
		av, inA := a[k]
		bv, inB := b[k]
		switch {
		case inA && !inB:
			ops = append(ops, patchOp{Op: "remove", Path: child})
		case !inA && inB:
			ops = append(ops, patchOp{Op: "add", Path: child, Value: bv})
		default:
			ops = append(ops, diffValue(child, av, bv)...)
		}
	}
	return ops
}

// keyedByID reports whether every element is an object with a string id.
func keyedByID(arr []any) bool {
	for _, item := range arr {
		obj, ok := item.(map[string]any)
		if !ok {
			return false
		}
		if _, ok := obj["id"].(string); !ok {
			return false
		}
	}
	return true
}

func diffKeyedArray(path string, a, b []any) []patchOp {
	var ops []patchOp

	aIndex := map[string]any{}
	aOrder := []string{}
	for _, item := range a {
		id := item.(map[string]any)["id"].(string)
		aIndex[id] = item
		aOrder = append(aOrder, id)
	}
	bIndex := map[string]any{}
	bOrder := []string{}
	for _, item := range b {
		id := item.(map[string]any)["id"].(string)
		bIndex[id] = item
		bOrder = append(bOrder, id)
	}

	// Removals first, in reverse index order, so earlier indices stay valid as the
	// patch is applied sequentially.
	for i := len(aOrder) - 1; i >= 0; i-- {
		if _, kept := bIndex[aOrder[i]]; !kept {
			ops = append(ops, patchOp{Op: "remove", Path: fmt.Sprintf("%s/%d", path, i)})
		}
	}

	// Modifications to surviving elements, addressed by their position in the new array.
	for newIdx, id := range bOrder {
		av, existed := aIndex[id]
		if !existed {
			continue
		}
		ops = append(ops, diffValue(fmt.Sprintf("%s/%d", path, newIdx), av, bIndex[id])...)
	}

	// Additions last, appended in order.
	for _, id := range bOrder {
		if _, existed := aIndex[id]; !existed {
			ops = append(ops, patchOp{Op: "add", Path: path + "/-", Value: bIndex[id]})
		}
	}

	return ops
}

// escapePointer applies RFC-6901 escaping.
func escapePointer(key string) string {
	key = strings.ReplaceAll(key, "~", "~0")
	return strings.ReplaceAll(key, "/", "~1")
}

func sameJSON(a, b any) bool {
	ea, errA := json.Marshal(a)
	eb, errB := json.Marshal(b)
	if errA != nil || errB != nil {
		return false
	}
	return string(ea) == string(eb)
}

func summarise(ops []patchOp) DiffSummary {
	s := DiffSummary{Paths: []string{}}
	for _, op := range ops {
		switch op.Op {
		case "add":
			s.Added++
		case "remove":
			s.Removed++
		case "replace":
			s.Replaced++
		case "move", "copy":
			s.Moved++
		}
		s.Paths = append(s.Paths, op.Path)
	}
	return s
}
